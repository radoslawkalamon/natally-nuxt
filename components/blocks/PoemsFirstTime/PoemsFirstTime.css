.poems-first-time {
  text-align: center;
}
