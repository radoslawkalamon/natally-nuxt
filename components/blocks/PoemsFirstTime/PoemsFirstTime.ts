import Vue from 'vue'
import type { VueConstructor } from 'vue'
import mixinPoemFirstTime from '@/utils/mixin.poem.firstTime'

export default (Vue as VueConstructor<
  Vue
  & InstanceType<typeof mixinPoemFirstTime>
>).extend({
  name: 'BlocksPoemsFirstTime',
  mixins: [mixinPoemFirstTime],
  computed: {
    shallShowSection (): boolean {
      return this['poem/firstTime/shallShow']
    }
  },
  mounted (): void {
    if (this.shallShowSection) {
      this['poem/firstTime/countVisit']()
    }
  }
})
