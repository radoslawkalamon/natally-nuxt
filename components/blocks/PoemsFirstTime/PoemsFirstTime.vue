<template>
  <client-only>
    <WrappersSection
      v-if="shallShowSection"
      v-once
      class="poems-first-time"
      data-test="blocks-poems-first-time"
    >
      <template #header>
        <ComponentsTitle
          :design="2"
          :type="2"
          title="Pierwszy raz?"
        />
      </template>

      <WrappersText>
        <p>
          Chcesz dowiedzieć&nbsp;się więcej o&nbsp;Poezji&nbsp;3.14?
        </p>
      </WrappersText>

      <template #footer>
        <ComponentsButtonLink
          label="Tak, z przyjemnością!"
          to="/poezja-314/"
        />
      </template>
    </WrappersSection>
  </client-only>
</template>

<script src="./PoemsFirstTime.ts" lang="ts"></script>
<style src="./PoemsFirstTime.css" scoped></style>
