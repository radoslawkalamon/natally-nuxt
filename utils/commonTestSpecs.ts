import Vue from 'vue'
import { mount } from '@vue/test-utils'
import { merge } from 'lodash'

export function shallRender (component: typeof Vue, dataTestID: string, defaultOptions?: object): void {
  test('shall render', (): void => {
    const wrapper = mount(component, defaultOptions)
    const isWrapperVisible = wrapper.find(`[data-test="${dataTestID}"]`).isVisible()
    expect(isWrapperVisible).toBeTruthy()
  })
}

export function shallHaveSlot (component: typeof Vue, slot: string, defaultOptions?: object): void {
  test(`shall have slot '${slot}'`, (): void => {
    const text = `This is slot '${slot}' test.`
    const wrapperOptions = merge(defaultOptions, {
      slots: {
        [slot]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHaveStringProp (component: typeof Vue, propName: string, defaultOptions?: object): void {
  test(`shall have string prop '${propName}'`, (): void => {
    const text = `This is prop ${propName} test.`
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHavePropInfluenceOverClassNames (
  component: typeof Vue,
  propName: string,
  propValue: unknown,
  classNameSuffix: string,
  defaultOptions?: object
) {
  test(`shall have ${typeof propValue} prop '${propName}' influence over class names`, () => {
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: propValue
      }
    })
    const wrapper = mount(component, wrapperOptions)
    const isWrapperHaveClass = wrapper.classes().some(cls => cls.includes(classNameSuffix))
    expect(isWrapperHaveClass).toBeTruthy()
  })
}
