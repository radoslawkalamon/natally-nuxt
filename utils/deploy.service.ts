/* eslint-disable no-console */
import { Client, FileInfo, FTPResponse } from 'basic-ftp'
import { ProgressInfo } from 'basic-ftp/dist/ProgressTracker'
import capitalize from './formatters/capitalize'
import spacelize from './formatters/spacelize'

type DeployServiceConfig = {
  host: string;
  port: number;
  username: string;
  password: string;
  remoteCatalog: string;
  remoteIgnoreFile: string[];
  localCatalog: string;
}

type RemoveStrategies = {
  [key in FileInfo['type']]: (name: FileInfo['name']) => Promise<void | FTPResponse>
}

function decoratorLogCall () {
  return function (_target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const targetMethod = descriptor.value

    descriptor.value = function (...args: any[]) {
      const formattedName = spacelize(capitalize(propertyKey))
      console.log(`::: ${formattedName} :::`)
      return targetMethod.apply(this, args)
    }

    return descriptor
  }
}

class DeployService {
  readonly host: string;
  readonly port: number;
  readonly username: string;
  readonly password: string;
  readonly remoteCatalog: string;
  readonly remoteIgnoreFile: string[];
  readonly localCatalog: string;
  FTPClient = new Client()

  constructor (config: DeployServiceConfig) {
    this.host = config.host
    this.port = Number(config.port)
    this.username = config.username
    this.password = config.password
    this.remoteCatalog = config.remoteCatalog
    this.remoteIgnoreFile = config.remoteIgnoreFile
    this.localCatalog = config.localCatalog
  }

  public async start (): Promise<void> {
    try {
      await this.openConnection()
      await this.ensureRemoteCatalog()
      await this.openRemoteCatalog()
      await this.cleanRemoteCatalog()
      await this.uploadToRemoteCatalog()
    } catch (e) {
      console.error(e)
      throw e
    } finally {
      this.closeConnection()
    }
  }

  @decoratorLogCall()
  private async openConnection (): Promise<void> {
    await this.FTPClient.access({
      host: this.host,
      port: this.port,
      user: this.username,
      password: this.password,
      secure: false
    })
  }

  @decoratorLogCall()
  private async ensureRemoteCatalog (): Promise<void> {
    await this.FTPClient.ensureDir(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async openRemoteCatalog (): Promise<void> {
    await this.FTPClient.cd(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async cleanRemoteCatalog (): Promise<void> {
    const remoteFiles = await this.FTPClient.list()

    for (const file of remoteFiles) {
      if (this.remoteIgnoreFile.includes(file.name)) {
        continue
      }

      const removeStrategies: RemoveStrategies = {
        0: async () => await Promise.resolve(),
        1: async name => await this.FTPClient.remove(name),
        2: async name => await this.FTPClient.removeDir(name),
        3: async () => await Promise.resolve()
      }
      await removeStrategies[file.type](file.name)
    }
  }

  @decoratorLogCall()
  private async uploadToRemoteCatalog (): Promise<void> {
    this.FTPClient.trackProgress(this.handlerUploadProgress)
    await this.FTPClient.uploadFromDir(this.localCatalog)
    this.FTPClient.trackProgress()
  }

  private handlerUploadProgress (info: ProgressInfo): void {
    if (info.bytes === 0) {
      process.stdout.write(`Uploading file ${info.name}\n`)
    }
  }

  @decoratorLogCall()
  private closeConnection (): void {
    this.FTPClient.close()
  }
}

export default DeployService
