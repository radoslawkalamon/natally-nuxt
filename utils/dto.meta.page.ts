export type DTOMetaPageConstructor = {
  createdAt: string;
  description: string;
  imageOpenGraph?: string;
  path: string;
  title: string;
  updatedAt: string;
}

const getPath = (path: string) => {
  const pathWithoutIndex = path.split('/').filter(i => i !== 'index').join('/')
  return `${pathWithoutIndex}${pathWithoutIndex.at(-1) !== '/' ? '/' : ''}`
}

export class DTOMetaPage {
  readonly createdAt: Date;
  readonly description: string;
  readonly imageOpenGraph: string;
  readonly path: string;
  readonly title: string;
  readonly updatedAt: Date;

  constructor ({
    createdAt,
    description,
    imageOpenGraph = 'common.opengraph.png',
    title,
    updatedAt,
    path
  }: DTOMetaPageConstructor) {
    this.createdAt = new Date(createdAt)
    this.description = description
    this.imageOpenGraph = `/images/${imageOpenGraph}`
    this.path = getPath(path)
    this.title = title
    this.updatedAt = new Date(updatedAt)
  }
}
