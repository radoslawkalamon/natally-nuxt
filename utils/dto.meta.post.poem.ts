import { DTOMetaPage, DTOMetaPageConstructor } from '@/utils/dto.meta.page'

export type DTOMetaPostPoemConstructor = DTOMetaPageConstructor & {
  imageCover: string
  imageOpenGraph: string
}

export class DTOMetaPostPoem extends DTOMetaPage {
  readonly imageCover: string;
  readonly imageOpenGraph: string;

  constructor ({
    createdAt,
    description,
    imageCover,
    imageOpenGraph,
    path,
    title,
    updatedAt
  }: DTOMetaPostPoemConstructor) {
    super({
      createdAt,
      description,
      path,
      title,
      updatedAt
    })

    this.imageCover = `${this.imagePath}${imageCover}`
    this.imageOpenGraph = `${this.imagePath}${imageOpenGraph}`
  }

  get imagePath (): string {
    return '/images/poezja-314/'
  }
}
