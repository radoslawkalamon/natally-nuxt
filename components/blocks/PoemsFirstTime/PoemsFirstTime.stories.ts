import { Meta, Story } from '@storybook/vue'
import PoemsFirstTime from './PoemsFirstTime.vue'

const meta: Meta = {
  title: 'Blocks / Poems First Time',
  component: PoemsFirstTime
}
export default meta

export const Default: Story = () => ({
  components: { PoemsFirstTime },
  template: '<PoemsFirstTime />'
})
