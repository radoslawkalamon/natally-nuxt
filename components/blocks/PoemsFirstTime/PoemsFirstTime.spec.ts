import PoemsFirstTime from './PoemsFirstTime.vue'
import { shallRender } from '@/utils/commonTestSpecs'

describe('Blocks / Poems First Time', () => {
  const defaultOptions = {
    stubs: {
      WrappersSection: {
        template: '<section data-stub="wrappers-section"><slot /></section>'
      },
      ComponentsTitle: {
        template: '<h1 data-stub="components-title" />'
      },
      WrappersText: {
        template: '<div data-stub="wrappers-text"><slot /></div>'
      },
      ComponentsButtonLink: {
        template: '<a href="#" data-stub="components-button-link" />'
      }
    }
  }

  shallRender(PoemsFirstTime, defaultOptions)
})
