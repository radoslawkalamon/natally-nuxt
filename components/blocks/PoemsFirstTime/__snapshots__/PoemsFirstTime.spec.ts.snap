// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blocks / Poems First Time Integration shall pass integration sanity test 1`] = `
"<section class=\\"section poems-first-time\\">
  <div data-test=\\"wrappers-section-header\\" class=\\"section__header\\">
    <h2 class=\\"title title--design-2 title--underscore\\">Pierwszy raz?</h2>
  </div>
  <div class=\\"section__inner\\">
    <div class=\\"text\\">
      <p>Chcesz dowiedzieć&nbsp;się więcej o&nbsp;Poezji&nbsp;3.14?</p>
    </div>
  </div>
  <div data-test=\\"wrappers-section-footer\\" class=\\"section__footer\\">
    <nuxtlink to=\\"/poezja-314/\\" class=\\"button-link\\">Tak, z przyjemnością!</nuxtlink>
  </div>
</section>"
`;

exports[`Blocks / Poems First Time Unit shall pass unit sanity test | hide 1`] = `""`;

exports[`Blocks / Poems First Time Unit shall pass unit sanity test | show 1`] = `
"<wrapperssection-stub class=\\"poems-first-time\\">
  <wrapperstext-stub>
    <p>Chcesz dowiedzieć&nbsp;się więcej o&nbsp;Poezji&nbsp;3.14?</p>
  </wrapperstext-stub>
</wrapperssection-stub>"
`;
